scala-package/macros/src/main/scala/org/apache/mxnet/APIDocGenerator.scala

```diff
@@ -203,26 +203,40 @@ private[mxnet] object APIDocGenerator{
   }
 
   def generateJavaAPISignature(func : absClassFunction) : String = {
+    val useParamObject = func.listOfArgs.count(arg => arg.isOptional) >= 2
     var argDef = ListBuffer[String]()
     var classDef = ListBuffer[String]()
     func.listOfArgs.foreach(absClassArg => {
       val currArgName = safetyNameCheck(absClassArg.argName)
       // scalastyle:off
-      if (absClassArg.isOptional) {
-        classDef += s"def set${absClassArg.argName}(${absClassArg.argName} : ${absClassArg.argType}) : ${func.name}BuilderBase"
+      if (absClassArg.isOptional && useParamObject) {
+        classDef += s"def set${absClassArg.argName.capitalize}(${absClassArg.argName}: ${absClassArg.argType}): ${func.name}ParamBase"
       }
       else {
         argDef += s"$currArgName : ${absClassArg.argType}"
       }
       // scalastyle:on
     })
-    classDef += s"def setout(out : NDArray) : ${func.name}BuilderBase"
-    classDef += s"def invoke() : org.apache.mxnet.javaapi.NDArrayFuncReturn"
     val experimentalTag = "@Experimental"
-    // scalastyle:off
-    var finalStr = s"$experimentalTag\ndef ${func.name} (${argDef.mkString(", ")}) : ${func.name}BuilderBase\n"
-    // scalastyle:on
-    finalStr += s"abstract class ${func.name}BuilderBase {\n  ${classDef.mkString("\n  ")}\n}"
+    val returnType = "org.apache.mxnet.javaapi.NDArrayFuncReturn"
+    var finalStr = ""
```

Comment:
You should try to combine all of these into as few string builders as possible as well as reduce extra variables. Just throw the classDef and argDef additions right into the string builder.

Reply:
Thanks for your feedback. It will happen once I start cherry-picking @mdespriee 's PR since there are big changes and optimization I can take from here. Just don't want to do it now as it would introduce too big change on the code base.